### ISSUE TITLE
Compare moodle `timemodified` with timestamp of local files

### ISSUE THREAD

**Issue by septatrix** [Romern/syncMyMoodle#77]
Sometimes course instructors upload newer versions of a file to change some mistakes etc.
This would be a relatively simple¹ way to catch those cases and download the newer version.

However if this is implemented we would also provide a solution for cases where all timestamps get updated (e.g. copying the files from one pc to another). For linux this could be done by a recursive `touch` to a day in the past however it would also be possible to use `os.utime` to set the timestamp to the one in Moodle. This could actually be done for every downloaded file

¹ In relation to always downloading every file and comparing the checksum.

**Comment by imjuzcy** [comment]
I was trying to implement this, but comparing the timestamp might not be efficient nor clean, as there are many reasons the timestamps might be changed without the content of the file changing. 

The other way would be to compare the file sizes. Although there are accurate to the byte, but there are also the very unlikely possibility that the content is changed without changing the size, especially for simple plain text files. 

The most accurate way is to keep a local copy of all metadata of the scraped files, and compare them on future scrapes. If the metadata is changed, redownload the file.

For reference, metadata from Moodle that we can compare:
`{'type': 'file', 'filename': 'ThisIsPDF.pdf', 'filepath': '/', 'filesize': 23526457, 'fileurl': 'https://moodle.rwth-aachen.de/webservice/pluginfile.php/412441/mod_folder/content/77/ThisIsPDF.pdf?forcedownload=1', 'timecreated': 1748255118, 'timemodified': 1748168741, 'sortorder': 0, 'mimetype': 'application/pdf', 'isexternalfile': False, 'userid': 412424, 'author': 'Authur, Author', 'license': 'allrightsreserved'}` from [this line](https://github.com/Romern/syncMyMoodle/blob/5ed31b8422d37e71719ead8b200e58bfbd224d09/syncmymoodle/__main__.py#L664)

**Comment by septatrix** [comment]
> I was trying to implement this, but comparing the timestamp might not be efficient nor clean, as there are many reasons the timestamps might be changed without the content of the file changing.

Yes but while it may be a bit overeager, it errs on the side of caution. Checking just the file size might miss updates as you pointed out. Comparing the whole metadata would also already include the timestamp so we might as well directly use that without the need to store auxiliary data.

One alternative would be to check if the server sends an ETag for the files themselves which we could check and store (along with the other Modified* headers)

**Comment by imjuzcy** [comment]
It does send an ETag, but then we will need to do a `HEAD` request for every file before download, which means a lot more requests.

Saving and comparing `timemodified` (in contrast to comparing with local file timestamp, which is platform dependant and unreliable) is probably the best way, all things considered. This way, there is no extra requests to the Moodle server than what is currently being done.

**Comment by septatrix** [comment]
> It does send an ETag, but then we will need to do a `HEAD` request for every file before download, which means a lot more requests.

No, we can simply send the `If-None-Match` header when requesting the file

> Saving and comparing `timemodified` (in contrast to comparing with local file timestamp, which is platform dependant and unreliable) is probably the best way, all things considered. This way, there is no extra requests to the Moodle server than what is currently being done.

I would not call the local file timestamp unreliable as long as one does not need ns precision (which we don't). It is available on all OSes and avoids the hassle of maintaining either a database or even worse, sidecar files. We would simple need to take care that after downloading a file, we set its `mtime` to the `timemodified` reported by Moodle. Then, we can assume that a file is always outdated if the `mtime` does not match the `timemodified`

**Comment by imjuzcy** [comment]
> No, we can simply send the If-None-Match header when requesting the file

That still means we have to request every file, even though we might not download the file at the end, right? 

> It is available on all OSes and avoids the hassle of maintaining either a database or even worse, sidecar files.

I was thinking to just expand the `root_node` object with the `timemodified` attribute from Moodle and just dump it, then load it on next run and use it to compare to the new `root_node`.
